cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -591,7 +591,7 @@ Status DictionaryEncode(FunctionContext* ctx, const Datum& value, Datum* out) {
       dict_chunks.emplace_back(std::make_shared<DictionaryArray>(
           dict_type, MakeArray(datum.array()), dictionary));
     }
-    *out = detail::WrapArraysLike(value, dict_chunks);
+    *out = detail::WrapArraysLike(value, func->out_type(), dict_chunks);
```

Comment:
I suppose this `out_type` is incorrect (it seems to be the type of the indices)

Reply:
@jorisvandenbossche you're correct. Also the hash kernel had a conflicting fix for the same basic problem from #5650 which should be resolved now

Reply:
Thanks for your thorough investigation